### PR TITLE
Enable parallel message processing by default + introduce configurable limit

### PR DESCRIPTION
**Pull request by cheenamalhotra**
Building upon work by @alanrenmsft to enable parallel message processing.. this PR enables it by default and also allows max limit to be user defined.

image

This feature has been enabled in dev environments for a long time and has not shown any significant side effects in past months. It has been adjusted to work in optimal way in PR https://github.com/microsoft/sqltoolsservice/pull/2216 to disable `SubsetRequests` to be part of this feature to prevent unordered results from surfacing.

But overall, the feature offers many performance benefits. In worst case scenario, users can always disable the setting.

_I would urge testers to keep an eye on any issues in the next week when doing release testing to report anything concerning and we can take a call to disable it for the release if needed._ /cc @kisantia

### CODE REVIEW

**Review by erinstellato-ms — COMMENTED** [review]
Well this is exciting!  

Users will like being able to set a max limit.  What is the default value, and do we have a "recommended" upper limit?  Or is there a max value that we can allow users to enter so they set it something which could freeze their system?

**Comment by cheenamalhotra** [comment]
> Users will like being able to set a max limit. What is the default value, and do we have a "recommended" upper limit? Or is there a max value that we can allow users to enter so they set it something which could freeze their system?

Default is 100, which should be optimal for most users. Max value alone can't do much to freeze the system, it's the number of 'active and slow' parallel operations that a user can trigger with that range that can cause system to freeze - which is not very likely to occur, but we'll have to study this out of our teams to find out :)

**Comment by erinstellato-ms** [comment]
> Default is 100, which should be optimal for most users. Max value alone can't do much to freeze the system, it's the number of 'active and slow' parallel operations that a user can trigger with that range that can cause system to freeze - which is not very likely to occur, but we'll have to study this out of our teams to find out :)

Noted.  @dlevy-msft this will definitely be something to make sure is documented as an option, and maybe some verbiage noting default of 100 and something around perf considerations (after teams have done some testing).

@cheenamalhotra do you have telemetry hooked up here at all that would be useful to either understand if it's been changed, what it's been changed to?  Also, what data will we be looking for where we would then recommend turning this off?

**Comment by cheenamalhotra** [comment]
> @cheenamalhotra do you have telemetry hooked up here at all that would be useful to either understand if it's been changed, what it's been changed to? Also, what data will we be looking for where we would then recommend turning this off?

Yes, added telemetry measurement.

We would be only recommending turning this off when there's an issue like https://github.com/microsoft/azuredatastudio/issues/24095 which would occur randomly. 

This issue is handled now and PMP is disabled for query batches, but any behavior that occurs due to unmanaged concurrency is when we would ask customer to disable this feature while we address the issue. Our design should be able to handle parallel requests in backend as much as possible since we have resource locking in place where it matters the most, e.g. connections, binding queue, etc. If there's an area that shows random issue due to this feature being enabled, we would either disable PMP for that request or fix it to make it concurrent.

**Review by kisantia — APPROVED** [review]

File: extensions/mssql/src/utils.ts [hunk 1]
```diff
@@ -155,12 +156,22 @@ export function setConfigPreloadDatabaseModel(enable: boolean): void {
...
+		return true; // default value
 	}
-	const setting = config.inspect(parallelMessageProcessingConfig);
-	return (azdata.env.quality === azdata.env.AppQuality.dev && setting?.globalValue === undefined && setting?.workspaceValue === undefined) ? true : config[parallelMessageProcessingConfig];
+	return config[parallelMessageProcessingConfig];
 }
 
+/**
+ * Retrieves configuration `mssql:parallelMessageProcessingLimit` from settings file.
+ * @returns true if setting is enabled in ADS or running ADS in dev mode.
```
Comment (Charles-Gagnon) [inline extensions/mssql/src/utils.ts:166]: update

File: extensions/mssql/src/utils.ts [hunk 2]
```diff
@@ -155,12 +156,22 @@ export function setConfigPreloadDatabaseModel(enable: boolean): void {
 export function getParallelMessageProcessingConfig(): boolean {
```
Comment (Charles-Gagnon) [inline extensions/mssql/src/utils.ts:156]: Update the return description

**Review by Charles-Gagnon — APPROVED** [review]

**Pull request merged**